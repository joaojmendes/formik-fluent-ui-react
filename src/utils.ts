import { FieldProps, getIn } from 'formik'

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>

export function getErrorMessage<T = any>({ field, form }: FieldProps<T>) {
  const error = getIn(form.errors, field.name)
  const touched = getIn(form.touched, field.name)

  return touched ? error : undefined
}

export function invokeAll(...callbacks: any[]) {
  return () => {
    for (const callback of callbacks) {
      if (callback && typeof callback === 'function') {
        callback()
      }
    }
  }
}

export function createFakeEvent({ name }: { name: string }) {
  return { target: { name } }
}
