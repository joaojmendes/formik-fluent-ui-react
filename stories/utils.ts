import { action } from '@storybook/addon-actions'
import { FormikActions } from 'formik'

export function handleSubmit<T>(
  values: T,
  { setSubmitting }: FormikActions<T>
) {
  action('submit:start')(values)
  setTimeout(() => {
    setSubmitting(false)
    action('submit:done')()
  }, 2000)
}
