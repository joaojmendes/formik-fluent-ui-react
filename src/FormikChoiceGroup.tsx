import { ChoiceGroup, IChoiceGroupProps } from '@fluentui/react'
import { FieldProps } from 'formik'
import * as React from 'react'
import { createFakeEvent, Omit } from './utils'

export function mapFieldToChoiceGroup<T = any>({
  form,
  field,
}: FieldProps<T>): Pick<
  IChoiceGroupProps,
  'selectedKey' | 'onChange' | 'name'
> {
  return {
    name: field.name,
    selectedKey: field.value,
    onChange: (_, option) => {
      form.setFieldValue(field.name, option ? option!.key : null)
      field.onBlur(createFakeEvent(field))
    },
  }
}

export type FormikChoiceGroupProps<T = any> = Omit<
  IChoiceGroupProps,
  'selectedKey' | 'name' | 'onChange' | 'onBlur' | 'form'
> &
  FieldProps<T>

export function FormikChoiceGroup<T = any>({
  field,
  form,
  ...props
}: FormikChoiceGroupProps<T>) {
  return <ChoiceGroup {...props} {...mapFieldToChoiceGroup({ field, form })} />
}
