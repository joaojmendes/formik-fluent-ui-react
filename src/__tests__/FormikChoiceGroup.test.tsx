import { Field, FieldProps, Form, Formik } from 'formik'
import { ChoiceGroup } from '@fluentui/react'
import * as React from 'react'
import renderer from 'react-test-renderer'
import { FormikChoiceGroup, mapFieldToChoiceGroup } from '../FormikChoiceGroup'
import { noop, serialize } from './utils'

class Values {
  public selection: string | number = 5
}

function createFieldProps(value: string | number = 6): FieldProps<string | number> {
  return {
    field: {
      value,
      onChange: jest.fn(),
      onBlur: jest.fn(),
      name: 'selection',
    },
    form: { setFieldValue: jest.fn(), handleBlur: jest.fn(() => jest.fn()) },
  } as any
}

const options = ['foo', 'bar', 'baz'].map(x => ({ key: x, text: x }))

test('<FormikChoiceGroup /> renders correctly as a field component', () => {
  const component = renderer.create(
    <Formik initialValues={new Values()} onSubmit={noop}>
      <Form>
        <Field
          name="selection"
          label="Selection"
          component={FormikChoiceGroup}
        />
      </Form>
    </Formik>
  )

  expect(component.toJSON()).toMatchSnapshot()
})

test('<FormikChoiceGroup /> renders a Fabric <ChoiceGroup />', () => {
  const fieldProps = createFieldProps()

  const formikChoiceGroup = renderer.create(
    <FormikChoiceGroup {...fieldProps} options={options} label="Selection" />
  )
  const fabricChoiceGroup = renderer.create(
    <ChoiceGroup
      {...mapFieldToChoiceGroup(fieldProps)}
      label="Selection"
      options={options}
    />
  )
  expect(serialize(formikChoiceGroup)).toBe(serialize(fabricChoiceGroup))
})

test('mapFieldToChoiceGroup() maps FieldProps to IChoiceGroupProps', () => {
  const { field, form, meta } = createFieldProps()
  const props = mapFieldToChoiceGroup({ form, field, meta })

  expect(props.selectedKey).toBe(field.value)

  props.onChange!(null as any, options[1])

  expect(form.setFieldValue).toHaveBeenCalledTimes(1)
  expect(form.setFieldValue).toHaveBeenCalledWith(field.name, options[1].key)

  // onChange also calls onBlur
  expect(field.onBlur).toHaveBeenCalled()
})
