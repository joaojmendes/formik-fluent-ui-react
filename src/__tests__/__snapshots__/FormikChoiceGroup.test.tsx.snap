// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormikChoiceGroup /> renders correctly as a field component 1`] = `
<form
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="applicationRole-47"
    role="application"
  >
    <div
      aria-labelledby="ChoiceGroup1-label"
      className="ms-ChoiceFieldGroup root-48"
      role="radiogroup"
    >
      <label
        className="ms-Label root-50"
        id="ChoiceGroup1-label"
      >
        Selection
      </label>
      <div
        className="ms-ChoiceFieldGroup-flexContainer flexContainer-49"
      />
    </div>
  </div>
</form>
`;
